File: TicTacToe.cpp
#include <iostream>
using namespace std;

// declare and initialize ROWS and COLUMNS
const int ROWS = 3;
const int COLS = 3;

int main()
{
    // function prototypes
    void printboard(char[][COLS]);
    void maketurn(char[][COLS], bool);
    bool didend(char[][COLS], bool);
   
    // declare 3x3 array and bool variable
    char board[ROWS][COLS];
    bool isplayerone = true;

    // create array with * as placeholders
    // nested for loop to go through each column of each row
    for (int i = 0; i < ROWS; ++i) {
        for (int s = 0; s < COLS; ++s) {
            board[i][s] = '*';
        }
    }

    // print board before playing
    printboard(board);

    while(true) {
        maketurn(board, isplayerone);
        printboard(board);
        // break only when the game ends
        if (didend(board, isplayerone) == true) {
            break;
        }
        // alternate between players as game continues
        if (isplayerone == true) {
            isplayerone = false;
        }
        else {
            isplayerone = true;
        }
    }
   
    return 0;
}

// printboard function that prints 3x3 array onto screen between moves
void printboard(char gameboard[][COLS])
{
    for (int i = 0; i < ROWS; ++i) {
        for (int s = 0; s < COLS; ++s) {
            cout << gameboard[i][s];
        }
        cout << endl;
    }

    return;
}

// maketurn function that allows for players to make a move as long as it is in parameters of the board (3x3) and is currently occupied by an *
void maketurn(char gameboard[][COLS], bool playeroneturn) 
{
    int playerrow, playercol;
    while (true) {
        if (playeroneturn == true) {
            cout << "Enter a row and column to place an X"  << endl;
        }
        else {
            cout << "Enter a row and column to place an O" << endl;
        }

        cout << "Row: " << endl;
        cin >> playerrow;
        cout << "Column: " << endl;
        cin >> playercol;

        //conditions so no illegal moves can be made
        if ((playerrow >= 1) && (playerrow <= ROWS) && (playercol >= 1) && (playercol <= COLS) && (gameboard[playerrow-1][playercol-1] == '*')) {
            break;
        }
        else {
            cout << "That location is not available. Select another location." << endl;
        }

    }
    
    // subtract one from playerrow and playercol since arrays start from 0 not 1 
    if (playeroneturn == true) {
        gameboard[playerrow-1][playercol-1] = 'X';
    }
    else {
        gameboard[playerrow-1][playercol-1] = 'O';
    }

    return;
}

// didend function that ends game after win or tie
bool didend(char gameboard[][COLS], bool playeroneturn) 
{
    char symbol;
    bool win;
    if (playeroneturn == true) {
        symbol = 'X'; 
    }
    else {
        symbol = 'O';
    }

    // for loop representing a horizontal win
    for (int i = 0; i < ROWS; ++i) {
        bool win = true;
        for (int s = 0; s < COLS; ++s) {
            if (gameboard[i][s] != symbol) {
                win = false;
            }
        }
        if (win == true) {
            if (playeroneturn == true) {
                cout << "Player one won";
            }
            else {
                cout << "Player two won";
            }
            
            return true;
        }
    }

    // for loop representing a vertical win
    for (int i = 0; i < COLS; ++i) {
        bool win = true;
        for (int s = 0; s < ROWS; ++s) {
            if (gameboard[s][i] != symbol) {
                win = false;
            }
        }
        if (win == true) {
            if (playeroneturn == true) {
                cout << "Player one won";
            }
            else {
                cout << "Player two won";
            }
            
            return true;
        }
    }

    // for loop representing a diagonal win from [0][0] to [2][2]
    bool diagonalwin = true;
    for (int i = 0; i < ROWS; ++i) {
        if (gameboard[i][i] != symbol) {
            diagonalwin = false;
        }
    }
    if (diagonalwin == true) {
        if (playeroneturn == true) {
            cout << "Player one won";
        }
        else {
            cout << "Player two won";
        }
            
        return true;
    }

    // for loop representing opposite diagonal win from [2][0] to [0][2]
    diagonalwin = true;
    for (int i = ROWS-1; i >= 0; --i) {
        if (gameboard[i][COLS-i-1] != symbol) {
            diagonalwin = false;
        }
    }
    if (diagonalwin == true) {
        if (playeroneturn == true) {
            cout << "Player one won";
        }
        else {
            cout << "Player two won";
        }
            
        return true;
    }

    // tie game as a result of no more * on the board
    bool tie = true;
    for (int i = 0; i < ROWS; ++i) {
        for (int s = 0; s < COLS; ++s) {
            if (gameboard[i][s] == '*') {
                tie = false;
            }
        }
    } 
    if (tie == true) {
        cout << "Players tied" << endl;

        return true;
    }

    return false;
}

File: tictactoe.py
# Define dimensions of the board
rows = 3
columns = 3

# Define a function that initializes an empty 3x3 matrix for the board
def initialize_board():
    board = []
    for i in range(rows):
        # Initialize 'empty' cells with *
        temp = ['*', '*', '*']
        board.append(temp)
    return board

# Define and call a function that prints the board
def print_board(board):
    for i in range(rows):
        print(board[i], '\n')

# Define a function for a player to make a move
def make_move(board, row, column, player):
    message = ''
    out_of_bounds_error = 'Out of bounds error, try again'              # Outside 3x3 grid
    occupation_error = 'This move has already been played, try again'   # Move already been made
    unknown_player_error = 'Unknown player number'                      # Player outside 1 and 2
    success_message = 'Move made successfully'                          # Successful move
    move_made = False

    if row > rows or row < 1 or column > columns or column < 1:
        message += out_of_bounds_error
    elif player != 1 and player != 2:
        message += unknown_player_error
    elif (board[row - 1][column - 1] != '*'):
        message += occupation_error
    else:
        if (player == 1):
            board[row - 1][column - 1] = 'X'
        elif (player == 2):
            board[row - 1][column - 1] = 'O'
        message += success_message
        move_made = True
        
    if move_made:
        print_board(board)

    print(message)
    return move_made

""" 
# Function testing
make_move(board, 3, 3, 1)
make_move(board, 2, 4, 1)
make_move(board, 0, 2, 1)
make_move(board, 3, 3, 3)
make_move(board, 3, 3, 2)
### Should output:
# * * * 
# * * * 
# * * X
# Move made successfully 
# Out of bounds error, try again
# Out of bounds error, try again
# Unknown player number
# This move has already been played, try again
"""

# Define a function to check if the board represents a winning game
def check_win(board, symbol):    
    # Diagonal wins (2 variations)
    diagonal_win_1 = all(board[i][i] == symbol for i in range(rows))
    diagonal_win_2 = all(board[i][columns - 1 - i] == symbol for i in range(rows))

    if diagonal_win_1 or diagonal_win_2:
        return True, symbol

    # Horizontal wins (3 variations)
    for s in range(rows):
        if all(board[s][j] == symbol for j in range(columns)):
            return True, symbol

    # Vertical wins (3 variations)
    for l in range(columns):
        if all(board[t][l] == symbol for t in range(rows)):
            return True, symbol

    # Tie game
    board_full = all(board[p][q] != '*' for p in range(rows) for q in range(columns))
    if board_full:
        return True, 0

    return False, None

# Define a function to output a message given a winner/tie
def final_message(winner):
    if winner == 0:
        print('Tie')
    elif winner == 'X':
        print('Player 1 Wins!')
    elif winner == 'O':
        print('Player 2 Wins!')

# Define a function to simulate the game 
def play_game():
    players = [1,2]
    symbols = ['X', 'O']
    curr = 0

    board = initialize_board()
    print_board(board)

    while(True):
        row = int(input("Input Row (1-3): "))
        column = int(input("Input Column (1-3): "))
        
        # If make_move returns false (move wasn't made), then repeat 
        while not make_move(board, row, column, players[curr]):
            row = int(input("Input Row: "))
            column = int(input("Input Column: "))
        
        # Check if the latest move won the game, if not move on the loop, if so break
        status, symbol = check_win(board, symbols[curr])
        if status:
            final_message(symbol)
            break

        # Alternate between players 
        curr = 1 - curr

# Uncomment the following line to run game
play_game()